src/providers/powervs/mod.rs

```diff
@@ -0,0 +1,386 @@
+            };
+            output.push(iface);
+        }
+
+        output.shrink_to_fit();
+        Ok(output)
+    }
+}
+
+impl MetadataProvider for PowerVSProvider {
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        let metadata = self.read_metadata()?;
+        Self::known_attributes(metadata)
+    }
+
+    fn hostname(&self) -> Result<Option<String>> {
+        let metadata = self.read_metadata()?;
+        let hostname = if metadata.local_hostname.is_empty() {
+            None
+        } else {
+            Some(metadata.local_hostname)
+        };
+        Ok(hostname)
+    }
+
+    fn ssh_keys(&self) -> Result<Vec<PublicKey>> {
+        self.fetch_publickeys()
+    }
+
+    fn networks(&self) -> Result<Vec<network::Interface>> {
+        let data = self.read_network_data()?;
+        let interfaces = Self::network_interfaces(data)?;
+        Ok(interfaces)
+    }
+
+    fn virtual_network_devices(&self) -> Result<Vec<network::VirtualNetDev>> {
+        warn!("virtual network devices metadata requested, but not supported on this platform");
+        Ok(vec![])
+    }
+
+    fn boot_checkin(&self) -> Result<()> {
+        warn!("boot check-in requested, but not supported on this platform");
+        Ok(())
+    }
+}
+
+impl Drop for PowerVSProvider {
+    fn drop(&mut self) {
+        if let Err(e) = crate::util::unmount(
+            self.temp_dir.path(),
+            3, // maximum retries
+        ) {
+            slog_scope::error!("failed to unmount powervs config-drive: {}", e);
+        };
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::io::Cursor;
+
+    #[test]
+    fn test_powervs_basic_attributes() {
+        let metadata = r#"
+{
+  "hostname": "test_instance-powervs.foo.cloud",
+  "name": "test_instance-powervs",
+  "uuid": "41b4fb82-ca29-11eb-b8bc-0242ac130003"
+}
+"#;
+
+        let bufrd = BufReader::new(Cursor::new(metadata));
+        let parsed = PowerVSProvider::parse_metadata(bufrd).unwrap();
+        assert_eq!(parsed.instance_id, "41b4fb82-ca29-11eb-b8bc-0242ac130003",);
+        assert_eq!(parsed.local_hostname, "test_instance-powervs",);
+
+        let attrs = PowerVSProvider::known_attributes(parsed).unwrap();
+        assert_eq!(attrs.len(), 2);
+        assert_eq!(
+            attrs.get("POWERVS_INSTANCE_ID"),
+            Some(&"41b4fb82-ca29-11eb-b8bc-0242ac130003".to_string())
+        );
+        assert_eq!(
+            attrs.get("POWERVS_LOCAL_HOSTNAME"),
+            Some(&"test_instance-powervs".to_string())
+        );
+    }
+
+    #[test]
+    fn test_powervs_parse_metadata_json() {
+        let fixture = File::open("./tests/fixtures/powervs/meta_data.json").unwrap();
+        let bufrd = BufReader::new(fixture);
+        let parsed = PowerVSProvider::parse_metadata(bufrd).unwrap();
+
+        assert!(!parsed.instance_id.is_empty());
+        assert!(!parsed.local_hostname.is_empty());
+        assert!(!parsed.public_keys.is_none());
```

Comment:
Can we have an explicit test for fetching public keys also?

Reply:
added.